### ISSUE TITLE
Support capsule types

### ISSUE THREAD

**Issue by LucasPickering** [martinohmann/hcl-rs#392]
First off, thanks so much for your work on this library. It's really thorough and well documented.

[The HCL spec allows for an additional value type called "capsules"](https://github.com/hashicorp/hcl/blob/main/spec.md#capsule-types). This is an opaque application-specific type. In my case I need to represent binary (i.e. non-UTF-8 bytes) data in my app, which is coincidentally the same example given in the spec. It'd be nice if this library supported that.

The best implementation I can think of is to add a generic param to `Value` (defaulted to `()`) and an additional variant to hold it. E.g.:

```rust
pub enum Value<Capsule = ()> {
    Null,
    Bool(bool),
    Number(Number),
    String(String),
    Array(Vec<Value>),
    Object(Map<String, Value>),
    Capsule(Capsule),
}
```

I've started implementing it via this method, [code is here if you want to look](https://github.com/LucasPickering/hcl-rs/tree/capsule). If you agree with this feature I'm happy to keep working on this and make a PR for it.

**Comment by martinohmann** [comment]
Hi there!

To me it's not really clear how capsule types are actually used, do you have some examples of your specific use case?

Would capsules be limited to the `Value` type only, or would we need to add these as a variant in the `Expression` enum too? If that is the case, my fear is that it would litter the generic `Capsule` parameter into almost every single HCL type (because all of them ultimately have an `Expression` somewhere nested inside them) unless we come up with something better that does not require the addition of generics.

One thing to solve this which comes to my mind could be a capsule type that's just something like

```rust
pub struct Capsule(Vec<u8>);
```

But I'm not sure if this would work for all use cases that one might have.

**Comment by LucasPickering** [comment]
@martinohmann Yes, with this solution we would have to add a `<Capsule>` param to almost every type :(. I've started doing that in my branch, but I haven't touched the eval code at all because for my app I'm reimplementing evaluation (I need some pretty custom logic for it).

A bit about my use case: I have terminal-based HTTP client called [Slumber](https://github.com/LucasPickering/slumber). Currently the configuration is written in YAML, but I'm investigating moving it to HCL. The HCL is used to define "recipes", which define how HTTP requests can be constructed. I need to represent binary data in HCL values because some HTTP requests contain non-textual data (e.g. user wants to upload an image file). The source of this binary data could be a file or shell command. For example:

```hcl
upload_image "request" {
  method         = "POST"
  url            = "${profile.host}/anything/image"
  body = {
    type = "form_multipart"
    data = {
      filename = "logo.png"
      image    = file({ path = "./static/slumber.png" })
    }
  }
}
```

In this case, the `file` function returns non-UTF-8 data, so there's no way to represent it in the base `Value` enum. So for my app I would use `Value<Vec<u8>>`, and the `file` function could return `Value::Capsule`.

Hopefully that makes sense without being too detailed. I totally get your concern about the type param infecting everything. My hope is that adding a default of `Capsule = ()` it doesn't add any cognitive load to the user, but even then it's still an eyesore in the code and tedious to work with.

I can definitely make this work in my app without any changes to hcl-rs, probably with something like:

```rust
pub enum RenderValue {
    Hcl(hcl::Value),
    Binary(Vec<u8>),
}
```

But this would only work for me because I'm reimplementing the evaluation logic anyway. For other use cases, that's probably not an option.

**Comment by martinohmann** [comment]
Wouldn't it already be workable if your `file` function produced a base64 encoded string of the binary data which you would decode in your application again? That would at least work around the issue of passing non-utf8 data through HCL.

**Comment by LucasPickering** [comment]
@martinohmann Yes that's an option. You would need to tag it specifically as binary data to differentiate from an actual string that happens to be base64, i.e. represent it as something like `{ binary = true, data = "..." }`. Capsules would be easier to work with though, since you can statically declare your capsule type and don't need to mess around with additional semantic rules. Capsules are also unambiguous, so you don't have to come up with a scheme to separate your special extension values from values that users can create.

Another option that avoids the generic param would be to use type erasure, i.e. store `Box<dyn Any>` in the `Capsule` variant:

```rust
pub enum Value {
    Null,
    Bool(bool),
    Number(Number),
    String(String),
    Array(Vec<Value>),
    Object(Map<String, Value>),
    Capsule(Box<dyn Any>),
}
```

This misses out on a lot of the benefits of having the generic param, such as being able to implement traits based on the param, and it requires the extra step of having to downcast everywhere you use it. But it is an option.

**Comment by martinohmann** [comment]
@LucasPickering Sorry for the silence, I've been busy with other stuff the last few weeks.

I did look into it a bit myself and I'd prefer a capsule implementation based on `Box<dyn Any>`. A generic implementation would mean changes to almost every single type and file and I'm currently not willing to pay this price for this relatively niche feature.

Sure it's not as nice in terms of usability as a generic implementation, but it would allow us to implement the feature with relatively few changes compared to the generic alternative. It should be a viable solution for the use case you have though. And maybe, if it turns out to be worth it, we can pick up the topic of a generic capsule type again at one point in the future.

We could also think about a wrapper type here instead of using `Box<dyn Any>` directly, e.g. something like this:

```rust
#[derive(Debug)]
pub struct Capsule(Box<dyn Any>);

impl Deref for Capsule {
    type Target = Any;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for Capsule {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl Capsule {
    pub fn into_inner(self) -> Box<dyn Any> {
        self.0
    }
}
```

What's your opinion on this?

**Comment by LucasPickering** [comment]
@martinohmann I think this is a good approach. I agree that adding the generic param everywhere would be really invasive. However, I've decided not to use HCL for my project for other reasons (I think it's just not the right choice for the user), so I don't have a need for this anymore. You're welcome to close this issue if you want, or we can leave it open in case any future users come along and want this feature. Thanks for your time on this regardless!

**Comment by martinohmann** [comment]
@LucasPickering Sure, no worries, I'll probably keep this open as a reminder. Maybe I'll take a stab at it myself at one point if time and motivation permits :).
